photon-core/src/main/java/org/photonvision/vision/processes/VisionSourceManager.java

```diff
@@ -62,24 +57,79 @@ public static VisionSourceManager getInstance() {
     }
 
-    VisionSourceManager() {}
-
-    public void registerTimedTask() {
-        TimedTaskManager.getInstance().addTask("VisionSourceManager", this::tryMatchCams, 3000);
+    VisionSourceManager() {
+        TimedTaskManager.getInstance().addTask("CameraDeviceExplorer", this::discoverNewDevices, 5000);
     }
 
-    public void registerLoadedConfigs(CameraConfiguration... configs) {
-        registerLoadedConfigs(Arrays.asList(configs));
+    void clearConfigState() {
+        configsLoaded.set(false);
     }
 
     /**
-     * Register new camera configs loaded from disk. This will add them to the list of configs to try
-     * to match, and also automatically spawn new vision processes as necessary.
+     * Register new camera configs loaded from disk. This will create vision modules for each camera
+     * config and start them.
      *
      * @param configs The loaded camera configs.
      */
     public void registerLoadedConfigs(Collection<CameraConfiguration> configs) {
-        unmatchedLoadedConfigs.addAll(configs);
+        if (!configsLoaded.compareAndSet(false, true)) {
+            logger.warn("Attempted to register loaded configs after they were already loaded");
+            return;
+        }
+        logger.info("Registering loaded camera configs");
+
+        configs.stream()
+                .map(VisionSourceManager::loadVisionSourceFromCamConfig)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .map(VisionModuleManager.getInstance()::addSource)
+                .forEach(module -> {
+                    var config = module.visionSource.cameraConfiguration;
+                    cameraInfoMap.put(config.uniqueName, CameraInfo.fromCameraConfig(config));
+                    module.start();
+                });
+
+        logger.info("Finished registering loaded camera configs");
+    }
+
+    public Optional<CameraConfiguration> configureNewVisionSource(String uniqueName) {
+        boolean alreadyUsed = VisionModuleManager.getInstance().getModules().stream()
+                .anyMatch(module -> module.visionSource.cameraConfiguration.uniqueName.equals(uniqueName));
+        var cfg = Optional.ofNullable(cameraInfoMap.get(uniqueName))
+            .filter(u -> !alreadyUsed)
+            .map(info -> createConfigForCameras(info, uniqueName));
+        cfg.flatMap(VisionSourceManager::loadVisionSourceFromCamConfig)
+            .map(VisionModuleManager.getInstance()::addSource)
+            .ifPresent(VisionModule::start);
+
+        return cfg;
+    }
+
+    protected void discoverNewDevices() {
+        if (!configsLoaded.get()) {
+            logger.debug("Not discovering new devices because configs are not loaded");
+            return;
+        }
+
+        List<CameraInfo> devices = filterAllowedDevices(getConnectedCameras(), Platform.getCurrentPlatform());
+
+        List<String> infoNames = cameraInfoMap.values().stream()
+            .map(CameraInfo::name)
```

Comment:
is `name` a unique identifier here?

Reply:
This is actually the area I was struggling with, how do I check what devices are *new* only by using the data accessible in a camera configuration? The reason it needs to be out of a camera configuration is due to how the loaded configs are passed in and I have to fill their uniqueName slots without actually ever polling the hardware.

Reply:
Maybe we just lock down by USB port? So if the USB port of this device is already in use? Or what am I forgetting